Informer Lifecycle Improvements

kevindelgado opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature


**What this PR does / why we need it**:
Implementation of Informer Lifecycle Managemt [design doc](https://docs.google.com/document/d/17QrTaxfIEUNOEAt61Za3Mu0M76x-iEkcmR51q0a5lis/edit?usp=sharing)
Discussed at the Nov 4 sig-api-machinery meeting:[video](https://youtu.be/Vc8V8xVxK3k?t=27)

It does a few things that are broken up by commit:
1. Adds the OnError method to the ResourceEventHandler interface so that the handler is notified of ListAndWatch errors. 
2. Adds the ability to define stop conditions when running individual informers via the new `RunWithStopOptions()` method on the `SharedInformer` interface. It exposes running informers with stop options on the various informer factories (currently only supports global stop options that apply to all informers of an informer factory)
3. Modifies the metadata informer factory and dynamic informer factory to start informers via `RunWithStopOptions()` and provides a way to retrieve the stop channel for given informer to determine when an individual informer is stopped. 
4. Refactors some CRD specific integration testing helpers out of the GC integration test into a shared util package so that RQ integration test can also use them without duplicating code.
5. Modifies GC controller to recognize when a CRD informer has stopped and removes the monitor for it.
6. Modifies RQ controller to recognize when a CRD informer has stopped and removes the monitor for it.
7. Modifies controller-manager to run the GC and RQ controllers with RunWithStopOptions (meaning it always stops an informer upon reflector ListAndWatch error).


This is tested manually by modifying controller-runtime to use the new interfaces. Also by running the resource quota controller and GC controller with the new interfaces to verify that the informer shuts down as expected.

Unit testing has been added to the relevant pieces of tools/cache and metadatainformer/dynamicinformer. Integration tests for both the garbage collector and quota controller test that the install CRD/uninstall CRD/reinstall CRD flow continues to work while the informer gets stopped and restarted as expected.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #79610
Unblocks work relating to https://github.com/kubernetes-sigs/controller-runtime/pull/1192

```release-note
1. Adds the OnError method to the ResourceEventHandler interface so that the handler is notified of ListAndWatch errors. 
2. Adds the ability to define stop conditions when running individual informers via the new `RunWithStopOptions()` method on the `SharedInformer` interface. It exposes running informers with stop options on the various informer factories (currently only supports global stop options that apply to all informers of an informer factory)
3. Modifies the metadata informer factory and dynamic informer factory to start informers via `RunWithStopOptions()` and provides a way to retrieve the stop channel for given informer to determine when an individual informer is stopped. 
4. Refactors some CRD specific integration testing helpers out of the GC integration test into a shared util package so that RQ integration test can also use them without duplicating code.
5. Modifies GC controller to recognize when a CRD informer has stopped and removes the monitor for it.
6. Modifies RQ controller to recognize when a CRD informer has stopped and removes the monitor for it.
7. Modifies controller-manager to run the GC and RQ controllers with RunWithStopOptions (meaning it always stops an informer upon reflector ListAndWatch error).
```

based on https://github.com/kubernetes/kubernetes/pull/98657

kevindelgado commented:
/assign @caesarxuchao 
/assign @DirectXMan12

fedebongio commented:
/triage accepted
/ok-to-test

fedebongio commented:
/assign @yliaog

kevindelgado commented:
/retest

kevindelgado commented:
/test pull-kubernetes-e2e-kind

kevindelgado commented:
/retest

DirectXMan12 commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -600,6 +665,39 @@ func (p *sharedProcessor) addListenerLocked(listener *processorListener) {
+	p.listeners = append(p.listeners[:index], p.listeners[index+1:]...)
+	for i := len(p.syncingListeners) - 1; i >= 0; i-- {
+		if p.syncingListeners[i].handler == handler {
+			p.syncingListeners = append(p.syncingListeners[:i], p.syncingListeners[i+1:]...)
```
oh, we're counting down here.

I'd just find the index, break, and then remove outside the loop -- it's gonna be clearer to folks reading this

DirectXMan12 replied:
also document the contract somewhere -- is it okay to add a handler twice?  why?  why not?

DirectXMan12 replied:
```go
target := -1
for i, listener := range p.syncingListeners {
  if listener.handler == handler {
    target = i
    break
  }
}

if target != -1 {
  p.syncingListeners = append(p.syncingListeners[:target], p.syncingListeners[target+1:]...)
}

kevindelgado replied:
Done. See https://github.com/kubernetes/kubernetes/pull/98536

DirectXMan12 commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -17,13 +17,14 @@ limitations under the License.
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
ditto

DirectXMan12 reviewed (changes requested):
first pass, did all the low-level stuff (modulo the last low-level ordering test)

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +190,120 @@ type SharedInformer interface {
+	// Err can only be called after h.Done() has closed otherwise it will return
+	// nil.
+	Err() error
+	// Close allows for stopping the ifnormer (and closing the done channel).
```
typo s/ifnormer/informer/

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +190,120 @@ type SharedInformer interface {
+// Err can only be called after h.Done() has closed otherwise it will return
+// nil.
+func (h *stopHandle) Err() error {
+	return h.err
```
what about adding the check for if h.Done() has closed?

kevindelgado replied:
custom stopHandle has been removed

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +190,120 @@ type SharedInformer interface {
+	h.err = err
+}
+
+// Close allows for stopping the ifnormer (and closing the done channel).
```
typo s/ifnormer/informer/

yliaog commented on staging/src/k8s.io/code-generator/_examples/HyphenGroup/informers/externalversions/factory.go:
```diff
@@ -165,10 +165,19 @@ func (f *sharedInformerFactory) InformerFor(obj runtime.Object, newFunc internal
 // API group versions.
 type SharedInformerFactory interface {
 	internalinterfaces.SharedInformerFactory
+	StartWithStopOptions(stopCh <-chan struct{})
```
same question, could it be unified with Start(...)?

kevindelgado replied:
done

DirectXMan12 commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -468,11 +468,13 @@ func GetAvailableResources(clientBuilder clientbuilder.ControllerClientBuilder)
 	metadataClient := metadata.NewForConfigOrDie(rootClientBuilder.ConfigOrDie("metadata-informers"))
-	metadataInformers := metadatainformer.NewSharedInformerFactory(metadataClient, ResyncPeriod(s)())
-
+	metadataInformers := metadatainformer.NewSharedInformerFactoryWithOptions(metadataClient, ResyncPeriod(s)(), metadatainformer.WithOnListError(onListErr))
```
fix `ResyncPeriod` so that it's just `ResyncPeriod(s)` not `ResyncPeriod(s)()`

DirectXMan12 replied:
(might as well so that we're consistent)

yliaog replied:
yeah, better to be consistent, i think

kevindelgado replied:
Upon second look, I don't think this is inconsistent like we thought (or at least how I thought).

I had thought that `defaultResync` was a `SharedInformerOption` that was passed as part of the multivariate options argument to the factory. It's actually a required argument that is just a duration. The reason it looks funky here (by returning a function that is called) is because it is supposed to generate a random duration anytime it gets invoked.

It's a little weird, but I don't think it's inconsistent (because it's not actually an option) and I'm not sure there's a way to change it without breaking existing interfaces. Let me know if you still think we should change it.

kevindelgado replied:
After speaking with Solly, we've agreed to leave it as is, I've left a not on `ResyncPeriod` to clarify why it returns a function instead of a duration.

DirectXMan12 commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -179,6 +180,11 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 		// Get the current resource list from discovery.
 		newResources := GetDeletableResources(discoveryClient)
 
+		// Filter resources that have been uninstalled and reinstalled stopped since the last sync.
```
reword this sentence, I'm having trouble parsing "uninstalled and reinstalled stopped"

kevindelgado replied:
done

DirectXMan12 commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -179,6 +180,11 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
+		// Filter resources that have been uninstalled and reinstalled stopped since the last sync.
+		// The following sync will recognize them as new resources and they will be restarted.
+		// If we do not filter them out, the informer that was previously stopped will not get restarted.
+		gc.dependencyGraphBuilder.filterStoppedResources(newResources)
```
could maybe just move this to be "old resources is a field on gc, and then when something gets stopped it gets removed from old resources", but eh, if this works it's prob fine as-is

kevindelgado replied:
The problem here is that `startMonitors` (where we listen for the doneCh to fire to indicate that a resource's informer has stopped) lives on the graph builder and can't add to a field that is on the gc. 

So I think without rearranging the code a bunch, it's simplest to just populate a set on the graph_builder and then the gc Sync will just consume that. This is what I've tried to do here, but let me know if you think there is an easier way.

DirectXMan12 commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -248,17 +253,21 @@ func (gb *GraphBuilder) startMonitors() {
 	<-gb.informersStarted
 
+	gb.sharedInformers.Start(gb.stopCh)
 	monitors := gb.monitors
```
what is this line for?

kevindelgado replied:
If I'm remembering our discussing, this line is needed to start informers that have been stopped (and are newly added) but is redundant for informers that are already running, but there isn't a clean way to avoid this redundancy. Is that right?

DirectXMan12 commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -248,17 +253,21 @@ func (gb *GraphBuilder) startMonitors() {
+			go func() {
+				<-doneCh
+				// push to set of recently stopped resources.
+				gb.stoppedResourcesLock.Lock()
```
nit, add a helper method for this

kevindelgado replied:
done

DirectXMan12 commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -248,17 +253,21 @@ func (gb *GraphBuilder) startMonitors() {
+		// TODO(kdelga): Should we only be doing this when ok?
+		if doneCh, ok := gb.sharedInformers.DoneChannelFor(gvr); ok {
+			go func() {
+				<-doneCh
```
comment on what happens if we happen to sync between receiving a signal on this channel and acquiring the lock

kevindelgado replied:
Comment added to the `addStoppedResources` method signature that the resource will wait until the following sync to restart its informer. Let me know if that is clear (and correct) or if you were thinking differently.

DirectXMan12 reviewed (changes requested):
few more comments

DirectXMan12 commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go:
```diff
@@ -248,6 +248,72 @@ func TestDynamicSharedInformerFactory(t *testing.T) {
+	}
+	testObject := newUnstructured("extensions/v1beta1", "Deployment", "ns-foo", "name-foo")
+	fakeClient := fake.NewSimpleDynamicClient(runtime.NewScheme(), []runtime.Object{testObject}...)
+	fakeClient.ForceListError()
```
```suggestion
	fakeClient.AddReactor("list", "deployment", testing.ReactionFunc(func(...) {
	   // something, something is actually important here
	   return something, something, fmt.Errorf("oh no!")
	})
```

(note, this might not be exactly the right syntax -- the reactors are pretty brittle)

https://pkg.go.dev/k8s.io/client-go/testing#ReactionFunc
https://pkg.go.dev/k8s.io/client-go/testing#Fake.AddReactor

kevindelgado replied:
Fixed

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -651,6 +665,39 @@ func (p *sharedProcessor) addListenerLocked(listener *processorListener) {
+	defer p.listenersLock.Unlock()
+
+	for i, listener := range p.listeners {
+		// This will panic if the handler is not comparable
```
this is shared informer, possibly used by many clients, panic would affect all clients even though only one client misused it. what about returning an err?

kevindelgado replied:
I will address in separate PR.

kevindelgado replied:
Done. See https://github.com/kubernetes/kubernetes/pull/98536/files#diff-10a340e966883e5205df63ad90ba6827f38f8a8d4b8614152c0baae9813a56c6R637

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -151,6 +151,12 @@ type SharedInformer interface {
 	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration)
+	// RemoveEventHandler removes an event handler from the shared informer.
+	// It returns true if the removal is successful
+	RemoveEventHandler(handler ResourceEventHandler) bool
```
better to return err, bool

kevindelgado replied:
I will address in separate PR.

kevindelgado replied:
Done. See https://github.com/kubernetes/kubernetes/pull/98536/files#diff-10a340e966883e5205df63ad90ba6827f38f8a8d4b8614152c0baae9813a56c6R637

yliaog commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -117,16 +117,6 @@ type GraphBuilder struct {
-// Run is intended to be called in a goroutine. Multiple calls of this is an
-// error.
-func (m *monitor) Run() {
-	m.controller.Run(m.stopCh)
```
m.controller.Run(...) is still needed to be called somewhere, no?

kevindelgado replied:
See comment below, this is actually a bug because the monitor's controller is actually a no-op.

yliaog replied:
i think it's better fix the bug in a separate PR first, instead of mixing the fix inside this PR, what do you think?

yliaog commented on pkg/controller/resourcequota/resource_quota_monitor.go:
```diff
@@ -268,7 +268,7 @@ func (qm *QuotaMonitor) StartMonitors() {
 			monitor.stopCh = make(chan struct{})
-			qm.informerFactory.Start(qm.stopCh)
+			qm.informerFactory.StartWithStopOptions(qm.stopCh)
 			go monitor.Run()
```
GC and RQ are using monitors similarly, not sure why you're treating them differently for GC and RQ

kevindelgado replied:
I'm confused, can you explain what you mean by "treating them differently" for GC and RQ?

yliaog commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -117,16 +117,6 @@ type GraphBuilder struct {
-// Run is intended to be called in a goroutine. Multiple calls of this is an
-// error.
-func (m *monitor) Run() {
-	m.controller.Run(m.stopCh)
```
m.controller.Run(...) still needs to be called, no?

kevindelgado replied:
No, this actually never did anything in the first place.

Previously, it was believed that calling sharedInformer.GetController() would return a real controller that could be ran. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/graph_builder.go#L177

but in reality, it just returns a dummy controller that does a no-op https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L452

Hence we needed to modify the informer to shut itself down (via the stopoptions) since there is no way to access the controller of an informer.

yliaog commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -82,6 +82,11 @@ type GraphBuilder struct {
 	// dependencyGraphBuilder
 	monitors    monitors
 	monitorLock sync.RWMutex
+	// stoppedResources is the set resources that have been stopped
```
s/is the set resources/is the set of resources/

yliaog replied:
"that have been stopped": stopped is hard to understand here, what does it mean to have resources stopped? should be the controllers for the resources that have been stopped, i think.

kevindelgado replied:
You're right, it's pretty unclear. I have edited it to try to make more sense. To your second point yes, stopped resources are resources for whom the shared informer (and underlying controller and reflector) have been stopped.

yliaog commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -120,6 +125,7 @@ type monitor struct {
 }
 
 type monitors map[schema.GroupVersionResource]*monitor
+type resourceSet map[schema.GroupVersionResource]struct{}
```
is it simpler to use bool instead of struct{}?

kevindelgado replied:
I am using struct{} to be consistent with the `ignoredResources` field on the GraphBuilder.

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/graph_builder.go#L113

RainbowMango replied:
> and the reason ignoredResources does it that way I think is just because empty structs are smaller than bools.

+1.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -233,40 +215,34 @@ var internalPackages = []string{"client-go/tools/cache/"}
 	klog.V(2).Infof("Stopping reflector %s (%s) from %s", r.expectedTypeName, r.resyncPeriod, r.name)
 }
 
 // Run calls RunWithStopOptions and only exits when stopCh is closed
```
the comment say "Run calls RunWithStopOptions", but the code does not, why not?

kevindelgado replied:
fixed, now it does.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -545,29 +438,27 @@ func (s *sharedIndexInformer) RunWithStopOptions(stopOptions StopOptions) {
+	s.controller.RunWithStopOptions(ctrlCtx, stopOptions)
 }
 
 func (s *sharedIndexInformer) Run(stopCh <-chan struct{}) {
```
could it call RunWithStopOptions(...)?

kevindelgado replied:
fixed with previous commit

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -575,6 +581,14 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 	}
 }
 
+func (s *sharedIndexInformer) RemoveEventHandler(handler ResourceEventHandler) bool {
```
i don't see it used in RQ or GC, is it going to be used?

kevindelgado replied:
I will address this in the separate PR (#98536) but the short answer is that controller-runtime uses the shared informer directly (rather than through a factory) and will use this functionality.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -217,32 +217,35 @@ var internalPackages = []string{"client-go/tools/cache/"}
 }
 
 // Run calls RunWithStopOptions and only exits when stopCh is closed
+// Deprecated: Use RunWithStopOptions instead.
```
have we agreed to deprecate it? it seems to have its own use case, and good to keep as a convenience wrapper around the RunWithStopOptions.

kevindelgado replied:
same answer as comment below.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -161,13 +162,13 @@ type SharedInformer interface {
 	RunWithStopOptions(ctx context.Context, stopOptions StopOptions)
+	// Run supports the legacy way interface for running shared index informers
+	// with just a stop channel.
+	// Deprecated: use RunWithStopOptions instead.
```
same question as that for reflector:
have we agreed to deprecate it? it seems to have its own use case, and good to keep as a convenience wrapper around the RunWithStopOptions.

kevindelgado replied:
Correct, we have not agreed to deprecate it, I have removed the comment. Initially I added it because there is no functional difference between calling Run and RunWithStopOptions (with stopOptions that do not shut the informer down), but the former is being kept in order to not break existing clients. I think it is useful to recommend new clients to use RunWithStopOptions, but you are correct that we have not agreed to deprecate it.

kevindelgado commented:
I split event handler removal and counting into a separate PR here https://github.com/kubernetes/kubernetes/pull/98536

It feels like a separate enough change that it doesn't seem right to attach to this PR even though they are both related to informer improvements we want in controller-runtime.

I will address feedback related to those changes in that PR.

kevindelgado commented on staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:
```diff
@@ -165,10 +165,19 @@ func (f *sharedInformerFactory) InformerFor(obj runtime.Object, newFunc internal
 
+// StartWithStoptions implements the controller-manager InformerFactory interface.
+// Because the sharedInformerFactory is used only with builtin types, there is no reason why StartWithStopOptions
+// should ever be used.
```
fixed

kevindelgado commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go:
```diff
@@ -248,6 +248,40 @@ func TestDynamicSharedInformerFactory(t *testing.T) {
+
+	gvr := schema.GroupVersionResource{Group: "extensions", Version: "v1beta1", Resource: "deployments"}
+	informerListerForGvr := target.ForResource(gvr)
+	target.StartWithStopOptions(ctx.Done())
```
done

kevindelgado commented on staging/src/k8s.io/client-go/informers/factory.go:
```diff
@@ -182,11 +182,21 @@ func (f *sharedInformerFactory) InformerFor(obj runtime.Object, newFunc internal
 type SharedInformerFactory interface {
 	internalinterfaces.SharedInformerFactory
 	ForResource(resource schema.GroupVersionResource) (GenericInformer, error)
+	StartWithStopOptions(stopCh <-chan struct{})
```
unified

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -121,20 +129,25 @@ func New(c *Config) Controller {
-// Run blocks; call via go.
-func (c *controller) Run(stopCh <-chan struct{}) {
+// RunWithStopOptions begins processing items, and will continue until stopOptions recognizes a stop condition.
+// RunWithStopOptions blocks; call via go.
```
See below for calling more than once.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -147,14 +160,26 @@ func (c *controller) Run(stopCh <-chan struct{}) {
 }
 
+// Run supports calling RunWithStopOptions with just a stop channel
+// that when closed, is the only stop condition that will stop the controller.
```
In the most common use case for the controller (ran by a shared informer that is started by a factory), it will never be called more than.

That being said, if the controller was being used as a library outside an informer factory, my understanding is that yes, controllers (and shared informers) can be ran more than once and that should be fine. 

I did a quick test to verify this and no issues arise, but it does create multiple reflectors that react separately to each event the reflector detects. I think this is how it’s meant to work because multiple consumers are likely to start their own copies of shared informers for a given resource.

Can you confirm @DirectXMan12

kevindelgado replied:
Also, just to be clear, nothing is changing about the behavior of `Run`, any concerns about it being called multiple times are the same now as they were before if I'm understanding your question correctly.

kevindelgado replied:
Oh I see. Sorry that deletion was a mistake, I added it back to Run (and RunWithStopOptions). The behavior on this should not be changed.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -154,9 +155,16 @@ type SharedInformer interface {
-	// The informer will be stopped when stopCh is closed.
+	// Run supports the legacy way interface for running shared index informers by
+	// merging the provided stopCh with the internal stopRead on the informer
+	// and calling RunWithStopOptions.
```
The previous comment has been moved to `RunWithStopOptions` as this is just to indicate that they should use that function instead for new use cases.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +190,120 @@ type SharedInformer interface {
+type DoneChannel <-chan struct{}
+
+// OnListErrorFunc is a type alias for function that takes an error
+// and determines whether or not to stop.
```
done

kevindelgado commented:
Thanks a bunch for the feedback @yliaog and @DirectXMan12! I've addressed all comments and believe it's ready for another review. 

I will have an update for the other PR with the EventHandler removal and count code up separately that will address some of the comments here that are relevant there.

mandelsoft committed:
support removal of event handlers from SharedIndexInformers

To be able to implement controllers that are dynamically deciding
on which resources to watch, it is required to get rid of
dedicated watches and event handlers again. This requires the
possibility to remove event handlers from SharedIndexInformers again.
Stopping an informer is not sufficient, because there might
be multiple controllers in a controller manager that independently
decide which resources to watch.

Unfortunately the ResourceEventHandler interface encourages to use
value objects for handlers (like the ResourceEventHandlerFuncs
struct, that uses value receivers to implement the interface).
Go does not support comparison of function pointers and therefore
the comparison of such structs is not possible, also. Such handlers
cannot be matched again after they have been added and therefore it
is only possible to remove handlers that are comparable.
Fortunately struct based handlers can also be passed by reference.
The user of the interface can therefore still use those
handlers and remove them again, by switching to a pointer argument.

The remove method checks whether a handler can be compared
and ignores uncomparable handlers in the removal process.
Removing of uncomparable handlers result in an error return.

Remark: If as the result of a handler removal a complete informer
should be disabled it is higly recommended to incorporate
pull request #98653, which fixes a race condition when stopping
watches for an informer using the stop channel.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -150,13 +151,23 @@ type SharedInformer interface {
 	// be competing load and scheduling noise.
 	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration)
+	// RemoveEventHandler removes an event handler from the shared informer.
+	// It returns true if the removal is successful
```
Done see https://github.com/kubernetes/kubernetes/pull/98536

yliaog commented on staging/src/k8s.io/client-go/metadata/metadatainformer/interface.go:
```diff
@@ -20,12 +20,15 @@ import (
 type SharedInformerFactory interface {
 	Start(stopCh <-chan struct{})
 	ForResource(gvr schema.GroupVersionResource) informers.GenericInformer
+	ForExistingStoppableResource(gvr schema.GroupVersionResource) (informers.GenericInformer, cache.DoneChannel, bool)
```
simmilar to pr #98657, this is breaking change to the interface. if people have concern with #98657, then this needs revisit too.

kevindelgado replied:
Added a note to the agenda for the next sig-api-machinery meeting

yliaog commented on staging/src/k8s.io/client-go/metadata/metadatainformer/interface.go:
```diff
@@ -20,12 +20,15 @@ import (
 	Start(stopCh <-chan struct{})
 	ForResource(gvr schema.GroupVersionResource) informers.GenericInformer
+	ForExistingStoppableResource(gvr schema.GroupVersionResource) (informers.GenericInformer, cache.DoneChannel, bool)
+	//DoneChannelFor(gvr schema.GroupVersionResource) (cache.DoneChannel, bool)
```
it is commented out, still needed?

yliaog replied:
remove?

yliaog commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go:
```diff
@@ -59,12 +44,98 @@ type dynamicSharedInformerFactory struct {
+//
+//	doneCh, ok := f.stoppableInformers[gvr]
+//	return doneCh, ok
+//}
```
commented out?

yliaog reviewed (commented):
thanks for the pr. like splitting out EventHandler removal change into its own pr, it's better to further split this pr into a few smaller pieces.

kevindelgado commented on staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:
```diff
@@ -58,12 +43,79 @@ type metadataSharedInformerFactory struct {
+	return doneCh, ok
+}
+
 func (f *metadataSharedInformerFactory) ForResource(gvr schema.GroupVersionResource) informers.GenericInformer {
```
Done via `ForStoppableResource`

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -96,12 +97,16 @@ type controller struct {
+	RunWithStopOptions(ctx context.Context, stopOptions StopOptions)
+
+	// Run is the legacy interface for running a controller that once called,
+	// can only be stopped by closing the stop channel.
```
See convo with Yu, not formally deprecating right now.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -146,9 +149,36 @@ func (c *controller) Run(stopCh <-chan struct{}) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer refCancel()
```
done

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller_test.go:
```diff
@@ -448,3 +449,33 @@ func TestPanicPropagated(t *testing.T) {
 	}
 }
+
+func TestControllerRunWithStopOptions(t *testing.T) {
```
done

kevindelgado commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -177,6 +183,21 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
+		delete(set, resource)
+		// filter the resource from gb's recentlyRemoved set
+		// so that it gets re-installed on the next sync cycle.
+		delete(gb.stoppedResources, resource)
```
I'm not sure what you mean, can you explain?

Once the resource is removed from the `newResources` set (on line 194) it should be deleted from the `stoppedResources` set so that on the next iteration of Sync `stoppedResources` is ONLY populated with resources that have been stopped since the last time this filter was called (on the previous sync). 

Let me know if there's a better way to structure this code.

yliaog replied:
1. so currently, the stopped resource needs to wait for one more sync cycle to get restarted, is it better to avoid the wait? it means gc.resyncMonitors(...) needs to be updated to understand the stopped & reinstalled resources in between syncs. it may be not worthwhile if the change is too complicated though.
2. "newResources = GetDeletableResources(discoveryClient)" is also called in the loop below the first call (at line 212 in garbagecollector.go), is filterStoppedResources(...) also needed there?
3. i was thinking to reset gb.stoppedResources outside the loop, but well, it works as is anyway.

kevindelgado replied:
1. We only know about newly installed resources when gc.Sync runs and checks discovery (via GetDeletableResources), so I'm not sure there's a straightforward way to restart reinstalled resources between syncs, but something, somwhere still needs to poll discovery.
2. Good catch, added there as well. Ideally, I think the filtering would be a part of GetDeletableResources, but that is not a method on the gc and it’s probably not worth modifying the interface right?
3. Ah I see, changed.

kevindelgado commented:
Please take another look @yliaog, I have addressed all of your comments. Most of the changes are minor, commentary, etc but the one change of substance was to change how the DoneChannel gets returned from `ForStoppableResource` as a result of this conversation with Solly https://github.com/kubernetes/kubernetes/pull/97214/files#r557786493

kevindelgado commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -468,11 +468,13 @@ func GetAvailableResources(clientBuilder clientbuilder.ControllerClientBuilder)
 	metadataClient := metadata.NewForConfigOrDie(rootClientBuilder.ConfigOrDie("metadata-informers"))
-	metadataInformers := metadatainformer.NewSharedInformerFactory(metadataClient, ResyncPeriod(s)())
-
+	metadataInformers := metadatainformer.NewSharedInformerFactoryWithOptions(metadataClient, ResyncPeriod(s)(), metadatainformer.WithStopOnListError(stopOnListErr))
```
typed informers are not ever going to be ran with stop options (because builtin resources cannot be removed from the cluster).

kevindelgado commented:
/retest

kevindelgado commented:
/retest

kevindelgado commented:
/retest

kevindelgado commented:
/test pull-kubernetes-e2e-kind

kevindelgado commented:
/retest

kevindelgado commented:
/retest

kevindelgado commented:
Ok @yliaog, all feedback addressed and I added the CRD integration tests for quota controller as well.

I have reorganized the commits so that each focuses on a single unit of the change here and no commits are dependent on successive commits.

Hopefully it is much easier to review as each commit can standalone as an independent PR (provided the proceeding commits have landed as they build off of one another).

The commit logs have more detail but basically they break down into:
1. Changes to the SharedInformer interface enabling run with stop options.
2. Changes to the informer factories enabling metadata/dynamic informer factories to start all informers with stop options (and a lot of autogenerated informer code to satisfy changes to the factory interface)
3. Extracting some CRD integration test specific helpers out of the gc test and into test/integration/util
4. Changing gc controller to track informers that are stopped for resources that are uninstalled.
5. Changing quota controller (same purpose as #4).
6. Changing the controller-manager’s informers to actually be stopped when the resource is uninstalled which enables the changes from 4 and 5 to come to fruition.

It should be more manageable now to review each commit individually but let me know if you think it can be broken down further. After mostly final review we can discuss if we want to split this up into two or more PRs.

kevindelgado commented:
/retest

kevindelgado commented:
/retest

kevindelgado commented:
/retest

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +186,21 @@ type SharedInformer interface {
+// StopOnListErrorFunc is a type alias for function that takes
+// an error from the reflector's List call and
+// and determines whether or not to stop the reflector (and corresponding controller and shared informer).
+type StopOnListErrorFunc func(error) bool
```
inside reflector, this function is called on the err returned from ListAndWatch, the err does not necessarily List error, it could be Watch error too. In this case, the StopOnListErrorFunc is a misnomer.

kevindelgado replied:
Yea I that is correct, but `StopOnListAndWatchErrorFunc` seemed like a mouthful. What do you think about `StopOnReflectorErrorFunc`? It's short, but isn't exactly descriptive if you're wondering why the reflector would error out. If you think `StopOnListAndWatchErrorFunc` is better I'm happy to go that route too, open to whatever you think the best name would be.

yliaog replied:
StopOnListErrorFunc is a bit misleading, StopOnListAndWatchErrorFunc is too long, what about StopOnErrorFunc? and add comment to explain that the error func is called inside reflector, and it is checked for both List and Watch errors.

kevindelgado replied:
Sounds good, change made and comment added.

kevindelgado replied:
changed throughout the PR

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +358,45 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+
+// TestSharedInformerRunWithStopOptions runs an informer with StopOnListError
+// set to always return true. It tests that when the underlying reflector does reach a
+// list error (upon trying to add an object) that the informer stopps running before a
```
s/stopps/stops/

kevindelgado replied:
done

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +358,45 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+			}
+		case <-time.After(time.Second):
+			if item.stopOnListError {
+				t.Errorf("shared informer SHOULD have stopped itself when stopOnListError is true, waited 3s")
```
it waited for 1s, not 3s

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +358,45 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+			if !item.stopOnListError {
+				t.Errorf("shared informer should NOT have stopped when stopOnListError is false")
+			}
+		case <-time.After(time.Second):
```
what about using ForeverTestTimeout (30s)? or you believe 1s is sufficiently long?

kevindelgado replied:
Same as my comment in `reflector_test.go`, it actually must wait this long in order to validate the stopOnListError=false case, so I've bumped it down to 200ms.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -147,14 +153,37 @@ func (c *controller) Run(stopCh <-chan struct{}) {
 
+// Run supports calling RunWithStopOptions with just a stop channel
+// that when closed, is the only stop condition that will stop the controller.
+// Deprecated: use RunWithStopOptions instead.
```
done

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller_test.go:
```diff
@@ -448,3 +449,38 @@ func TestPanicPropagated(t *testing.T) {
+	select {
+	case <-ctx.Done():
+		break
+	case <-time.After(5 * time.Second):
```
I realized this test was not testing the stopOnListError=false case, so I've modified it to be consistent with reflector_test and shared_informer_test.

See my comment in `reflector_test.go`, it actually must wait this long in order to validate the stopOnListError=false case, so I've bumped it down to 200ms.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -212,19 +212,41 @@ func (r *Reflector) setExpectedType(expectedType interface{}) {
 			r.watchErrorHandler(r, err)
+			// default to never stopping (same as previous behavior that only had stop channel)
+			if onListErr := stopOptions.StopOnListError; onListErr != nil {
+				if stopOptions.StopOnListError(err) {
```
done

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -212,19 +212,41 @@ func (r *Reflector) setExpectedType(expectedType interface{}) {
+			// default to never stopping (same as previous behavior that only had stop channel)
+			if onListErr := stopOptions.StopOnListError; onListErr != nil {
+				if stopOptions.StopOnListError(err) {
+					klog.V(2).Infof("Closing with list err, type %s", r.expectedTypeName)
```
done

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/reflector_test.go:
```diff
@@ -105,6 +106,54 @@ func TestRunUntil(t *testing.T) {
+			if !item.stopOnListError {
+				t.Errorf("reflector should NOT have stopped when stopOnListError is false")
+			}
+		case <-time.After(3 * time.Second):
```
I'm not sure we want to use that here, because we actually are going to wait this long for the test to pass. We are testing both when stopOnListError is true and false, so in the false condition, it will not stop and we expect it to timeout after the desired wait time. I'm open to better suggestions about what to use as the wait time, it just needs to be long enough to not fire before stopOnListError does. 

I tried injecting a fake clock into the reflector, but the issue is that you still need to block some amount of time in order for the go routine running the reflector to start it and recognize the list error. 

I actually think it can be much less than 3 seconds though, so in order to avoid slowing down the test too much, I've bumped it down to 200ms for both here and in the shared_informer test. It really only needs to wait for as long as it takes the reflector to make its inital list call and error out.

Let me know if you think there is a better approach.

yliaog replied:
the timeout cannot be too short, but it can be arbitrarily long, right? i think the best way is to run the test 10000 times under stress, varying timeout to see what is the best timeout value. i think a few seconds are fine, the tests are run concurrently, the timeout will slow down this one test, but it won't affect other tests.

go test k8s.io/client-go/tools/cache --race --count 10000
or use https://pkg.go.dev/golang.org/x/tools/cmd/stress

kevindelgado replied:
Cool, thanks for the advice. I ran it with stress and started seeing about 1 failure every couple thousand runs at 10ms. 20ms hit an error at 16k runs, and 100ms seemed to go fine fora as long as I let the stress test run for (100k+ tests).

I've added a comment explaining this.

If you'd feel more comfortable with a higher wait I can bump it up higher too.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -154,8 +155,11 @@ type SharedInformer interface {
+	// The informer will be stopped when a stopOptions condition is met.
+	RunWithStopOptions(ctx context.Context, stopOptions StopOptions)
+	// Run supports the legacy way interface for running shared index informers
+	// with just a stop channel. Use RunWithStopOptions.
```
done

kevindelgado commented:
@yliaog feedback addressed and changes squashed into the first `tools/cache` pkg commit, PTAL

kevindelgado commented:
/retest

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/controller_test.go:
```diff
@@ -448,3 +449,51 @@ func TestPanicPropagated(t *testing.T) {
+// before cancelling the context. If the context is stopped before the test
+// timeout, that means the controller stopped itself as expected.
+func TestControllerRunWithStopOptions(t *testing.T) {
+	waitTime := 200 * time.Millisecond
```
done

RainbowMango commented on staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:
```diff
@@ -58,12 +43,95 @@ type metadataSharedInformerFactory struct {
 
 var _ SharedInformerFactory = &metadataSharedInformerFactory{}
 
+// WithStopOnError sets the stopOnListErrFunc that is added to the stop options
```
```suggestion
// WithStopOnError sets the stopOnErrorFunc that is added to the stop options
```
A typo?

kevindelgado replied:
Yep. Fixed here and in the dynamicinformer, thanks!

RainbowMango commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -166,15 +166,30 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
+		// so that it's deletion is recognized on the first Sync cycle.
+		delete(set, resource)
+	}
+	// reset stoppedResources so that all stopped resources
+	// are re-installed on the next sync cycle.
```
Do you mean the informer of stopped resources,  that are re-installed on the same sync cycle, will be re-start on the next sync cycle?

kevindelgado replied:
Yes, comment updated

RainbowMango commented on staging/src/k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions/factory.go:
```diff
@@ -169,6 +169,7 @@ func (f *sharedInformerFactory) InformerFor(obj runtime.Object, newFunc internal
 // API group versions.
 type SharedInformerFactory interface {
 	internalinterfaces.SharedInformerFactory
+	ForStoppableResource(resource schema.GroupVersionResource) (*StoppableInformerInfo, bool)
```
Is the second `bool` return value redundant? If `*StoppableInformerInfo` == nil it implies not exist.

Another question is, from the way of function naming, `ForStoppableResource` and `ForResource` are similar, they should be no dependencies with each other. What about give  `ForStoppableResource` a new name, such as `GetStoppableInformer`, or call `ForResource` automatically?

kevindelgado replied:
Yea, I think originally there was some confusion around a situation where the informer existed but was not started. I have modified `ForStoppableResource` to call `ForResource` initially. This should make the naming choice more sensible and got rid of the bool.

RainbowMango commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go:
```diff
@@ -59,12 +44,95 @@ type dynamicSharedInformerFactory struct {
 	// This allows Start() to be called multiple times safely.
-	startedInformers map[schema.GroupVersionResource]bool
-	tweakListOptions TweakListOptionsFunc
+	startedInformers   map[schema.GroupVersionResource]bool
+	stoppableInformers map[schema.GroupVersionResource]cache.DoneChannel
```
Just a question, the `startedInformers` and `stoppableInformers` always hold the same GVR set, right?

kevindelgado replied:
Yep. This is a vestige from when I had separate `Start` and `StartWithStopOptions` methods. Changed so that `startedInformers` is a map to the done channel and removed `stoppableInformers`

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +186,21 @@ type SharedInformer interface {
+type DoneChannel <-chan struct{}
+
+// StopOnErrorFunc is a type alias for function that receives an error
+// anytime time the underlying reflector's ListAndWatch call returns an error
```
s/anytime time/anytime/

kevindelgado replied:
done

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -182,6 +186,21 @@ type SharedInformer interface {
+
+// StopOptions let the caller specify which conditions to stop the informer.
+type StopOptions struct {
+	// StopOnError inspects errors returned from the underlying reflector's List call,
```
s/List call/ListAndWatch call/

kevindelgado replied:
done

yliaog commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go:
```diff
@@ -248,6 +250,65 @@ func TestDynamicSharedInformerFactory(t *testing.T) {
+		onErrorFunc := func(error) bool {
+			return ts.stopOnError
+		}
+		testObject := newUnstructured("extensions/v1beta1", "Deployment", "ns-foo", "name-foo")
```
prefer apps/v1

yliaog commented on staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go:
```diff
@@ -248,6 +250,65 @@ func TestDynamicSharedInformerFactory(t *testing.T) {
+		}
+		testObject := newUnstructured("extensions/v1beta1", "Deployment", "ns-foo", "name-foo")
+		gvrToListKind := map[schema.GroupVersionResource]string{
+			{Group: "extensions", Version: "v1beta1", Resource: "deployments"}: "DeploymentList",
```
better to use apps/v1

yliaog commented on test/integration/util/extension.go:
```diff
@@ -0,0 +1,163 @@
+	t *testing.T, apiExtensionClient apiextensionsclientset.Interface,
+	dynamicClient dynamic.Interface,
+	namespace string,
+) (*apiextensionsv1beta1.CustomResourceDefinition, dynamic.ResourceInterface) {
```
this formatting looks weird, could this line be merged to prior line?

kevindelgado replied:
done

yliaog commented on test/integration/util/extension.go:
```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
s/2018/2021/? how much have you changed? it's hard to see the diff in a new file.

kevindelgado replied:
done. I've changed very little, but I guess there's no harm in keeping the copyright up to date.

yliaog commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -161,6 +161,9 @@ controller, and serviceaccounts controller.`,
 // hammer the apiserver with list requests simultaneously.
+// This returns a function rather than just a duration because it is stored
+// on the ControllerContext to be invoked multiple times.
+// If you just need to use it once, immediately invoke it.
```
the added comment seems to repeat the original comment with little extra info. consider merging it into the original comment.

kevindelgado replied:
done

yliaog commented on test/integration/garbagecollector/garbage_collector_test.go:
```diff
@@ -216,106 +176,49 @@ func setup(t *testing.T, workerCount int) *testContext {
-	}
-}
-
 func TestCrossNamespaceReferencesWithWatchCache(t *testing.T) {
```
could you add a test case or two to test?
1. CRD is installed, things working as expected 
2. then uninstalled, things working as expected
3. then reinstalled, things working as expected

kevindelgado replied:
Take a look at the existing `TestCRDDeletionCascading`. If I'm understanding you correctly, that is exactly what this test is already doing. It has been the primary test I've used to confirm my GC changes do/don't work and it:
1. Installs a CRD
2. Creates a CR with a dependent on a builtin type
3. Uninstalls the CRD
4. Confirms that both the CR and dependent are deleted
5. Reinstalls the CRD and repeats steps 2-4.

kevindelgado replied:
I've added a check to assert that the CRD informer must stop in a reasonable amount of time after CRD uninstall. 

https://github.com/kubernetes/kubernetes/pull/97214/files#diff-869f9a4b37327530c4d7890e790937adc33b0033dbad0a1ea0739f121f971d31R1175-R1182

yliaog commented on pkg/controller/resourcequota/resource_quota_monitor.go:
```diff
@@ -263,17 +242,16 @@ func (qm *QuotaMonitor) StartMonitors() {
+		if info := qm.informerFactory.ForStoppableResource(gvr); info != nil {
+			go func() {
+				<-info.Done
+				delete(qm.monitors, gvr)
```
could there be race without lock?

kevindelgado replied:
yes, good catch. Done.

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+	dynamicClient      dynamic.Interface
+	metadataClient     metadata.Interface
+	startRQ            func(workers int)
+	// syncPeriod is how often the GC started with startGC will be resynced.
```
s/GC/RQ/?

kevindelgado replied:
done throughout

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+	syncPeriod time.Duration
+}
+
+// if workerCount > 0, will start the GC, otherwise it's up to the caller to Run() the GC.
```
s/GC/RQ/

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+func setupWithServer(t *testing.T, result *kubeapiservertesting.TestServer, workerCount int) *testContext {
+	ctx := util.ExtensionSetup(t, result, workerCount)
+
+	// TODO: refactor with extensionSetup
```
remove the TODO?

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+		_, err := resourceClient.Create(context.TODO(), util.NewCRDInstance(definition, ns.Name, fmt.Sprintf("crd-%d", i)), metav1.CreateOptions{})
+		if err != nil {
+			if i != n+1 {
+				t.Fatalf("failed to create crd on iteration: %d, err: %v", i, err)
```
s/failed to created crd/failed to create crd instance/

kevindelgado replied:
done

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+			// equals the number of crd instances created
+			err := pollExpectedQuota(ctx.clientSet, i, resourceName, quotaName, ns.Name)
+			if err != nil {
+				t.Fatalf("err %v", err)
```
better to add extra descriptions to the err

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -88,10 +100,13 @@ func TestQuota(t *testing.T) {
 
-	informers := informers.NewSharedInformerFactory(clientset, controller.NoResyncPeriodFunc())
+	metadataClient := metadata.NewForConfigOrDie(clientConfig)
+	stopOnError := func(error) bool { return true }
```
better to have another test case for stopOnError return false

kevindelgado replied:
I can do this if you think it's the right thing to do, but I just want to point out that we never expect quota and GC controllers to ever be ran with stopOnError to be false. This is an error state and is what causes the bug in #79610 in the first place. I think it would be confusing to have tests for a state we consider to be an error in the first place (even if the controller techinically does still work because the informer never stops).

The only reason we don't change the informer/reflector to always stop on error is because this could break custom CRD controllers that we have no visibility in.

If you do think it's useful to test this state, I can do so here as well as throughout the GC tests, but it will double the number of tests runs in the GC and quota `test/integration` packages because every test runs with metadata informers and so if we do want to test this state we'd probably want to run every test with these states.

kevindelgado replied:
Same solution as GC test, I've added a check to confirm that the informer is stopped upon CRD uninstall:

https://github.com/kubernetes/kubernetes/pull/97214/files#diff-f9a432a9226aae7f52a9e21d3dbcc354a259a530eba2e2e11c97e0b760f4b2f5R433-R439

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+		t.Fatalf("failed to create crd instance 1, err: %v", err)
+	}
+
+	// confirm crd shows up in quota used count
```
s/crd/crd instance/ or just 'cr'

yliaog commented on test/integration/quota/quota_test.go:
```diff
@@ -371,3 +388,208 @@ func TestQuotaLimitedResourceDenial(t *testing.T) {
+		t.Fatalf("err %v", err)
+	}
+
+	// delete the definition which should cascade to the instance and quota?
```
no, quota should not be deleted, it is just not effective anymore

kevindelgado replied:
fixed

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -179,6 +180,11 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
+		// Filter resources that have been uninstalled and reinstalled on the cluster since the last sync.
+		// The following sync will recognize them as new resources and they will be restarted.
+		// If we do not filter them out, the informer that was previously stopped will not get restarted.
+		gc.dependencyGraphBuilder.filterStoppedResources(newResources)
```
Do I understand this correctly: The uninstalled CRDs are filtered out, so even if they have been reinstalled before this sync, they won't be restarted in this sync, and they need to wait for the next sync.

kevindelgado replied:
yes that is correct

yliaog replied:
@caesarxuchao could you help review the bug fix in GC? in case i miss anything there.

kevindelgado commented:
/retest

kevindelgado reviewed (commented):
I've added checks to both the GC and RQ integration tests to confirm that the informer is stopped when the CRD is uninstalled.

This will fail if stopOnError is false and answers the question of "how do we test that the new changes are in effect if the controller succeeds regardless of whether the informer stops or not"

PTAL @yliaog

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller_test.go:
```diff
@@ -448,3 +449,57 @@ func TestPanicPropagated(t *testing.T) {
+	// It needs to just be long enough for the underlying reflector to make
+	// its initial list call and error out.
+	//
+	// We begin seeing flakes every couple thousand runs at 10ms, so 100ms
+	// was chosen as an order of magnitude higher and does not seem to ever flake.
```
is this true even when running under [stress](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md#deflaking-unit-tests)? CI CPU is significantly lower than dev machines

kevindelgado replied:
yup, stress is what I used to test this, but see below, happy to bump it up if it's a concern.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller_test.go:
```diff
@@ -448,3 +449,57 @@ func TestPanicPropagated(t *testing.T) {
+			if !item.stopOnError {
+				t.Errorf("controller should NOT have stopped when stopOnError is false")
+			}
+		case <-time.After(waitTime):
```
waits less than a second (or few) very frequently cause flakes in resource-constrained environments like CI

kevindelgado replied:
Yea, Yu had the same concerns. I was open to bumping it further, but because the stress results seem to indicate things only got hairy around 10ms, it seemed like leaving it at 100ms was fine. 

But since you're the second (of two) reviewer to express this concern, I feel like I should just bump it to a second and call it a day. Let me know if you want it bumped higher to two or three seconds.

liggitt commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -488,8 +490,9 @@ func CreateControllerContext(s *config.CompletedConfig, rootClientBuilder, clien
 	versionedClient := rootClientBuilder.ClientOrDie("shared-informers")
 	sharedInformers := informers.NewSharedInformerFactory(versionedClient, ResyncPeriod(s)())
 
+	stopOnError := func(error) bool { return true }
```
does this stop all informers used by anything in kube-controller-manager if there's a single list or watch error? what restarts them?

kevindelgado replied:
No, it only stops metadata informers on a list or watch error. The typed informer factory continues to start informers via `Run` rather than `RunWithStopOptions` and thus these informers never get shut down.

kevindelgado replied:
It does feel a little weird and non-obvious to have typed informers completely ignore the stopOnError, but I'm not sure there's a better work around given the controller-manager informer factory is awrapper around both typed informer factory and metadata informer factory.

kevindelgado replied:
Let me know if you have any suggestions

liggitt commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -179,6 +180,11 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
+		// Filter resources that have been uninstalled and reinstalled on the cluster since the last sync.
+		// The following sync will recognize them as new resources and they will be restarted.
+		// If we do not filter them out, the informer that was previously stopped will not get restarted.
+		gc.dependencyGraphBuilder.filterStoppedResources(newResources)
```
this mutates both the input variable and `dependencyGraphBuilder` state in a way that I find very hard to reason about. (it also happens before we've done sanity checks on `newResources` to make sure we even got results and that they differ from our last check)

I expected no-longer-existing resources to be handled in the guts of `syncMonitors` (where we're currently calculating `toRemove` already based on delta from the last set of resources we started monitors for)

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -82,6 +82,21 @@ type GraphBuilder struct {
+	// from the cluster in between gc Syncs.
+	// When the resource is uninstalled the shared informer for the resource
+	// (as well as the underlying controller and reflector) are stopped.
+	// We need to track these resources because if they are reinstalled before
+	// the next Sync, the garbage collector will be unaware that the informer
+	// has been stopped and will not restart the informer.
```
this bit doesn't seem right... syncMonitors should always leave us with running monitors for the passed-in resources, right?

kevindelgado replied:
fixed, see above explanation

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -166,15 +166,31 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
+	return generic.Informer(), nil
+}
+
+// filterStoppedResources filters out the resources that have been removed
+// and reinstalled from the cluster in between garbage collector syncs.
+func (gb *GraphBuilder) filterStoppedResources(set resourceSet) {
```
mentioned above, but echoed here: this mutates both the input variable and `gb` state in a way that I find very hard to reason about

kevindelgado replied:
fixed, see above explanation

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -210,27 +226,33 @@ func (gb *GraphBuilder) syncMonitors(resources map[schema.GroupVersionResource]s
 	}
 	gb.monitors = current
 
-	for _, monitor := range toRemove {
```
This seems like it would have been an ideal place to explicitly stop the started informer based on no longer wanting to watch this resource. Is there a reason we can't be more direct with stopping informers for resources no longer listed in discovery?

kevindelgado replied:
fixed, see above explanation

I think doing it in `startMonitors` is sufficient and prevents an unnecessary indirection

kevindelgado reviewed (commented):
Thanks for the review. I think I've addressed all feedback. The gc stopped resource filtering has been a sticking point for a while, but looking at it again made me see where the issue was.

PTAL @liggitt

kevindelgado commented:
/retest

kevindelgado commented:
Friendly ping @liggitt @yliaog

yliaog commented:
thanks for the pr. it looks good, please have liggitt to take another look.
/lgtm

liggitt commented:
Synced offline. I think the way this PR is adding the ability to stop informers leads to unsafe failure modes.

The decision/policy to stop an informer based on an error in this PR is done top-level as part of starting the SharedInformerFactory. This is disconnected from the consumers of the Informers, which are registered/scattered throughout the codebase.

This approach relies on sweeping the codebase to ensure all informer consumers expect/handle their informers stopping, and reregister as needed if they want their informer to resume. This PR happened to do that for GC and Quota controllers, who are the current consumers of the metadata sharedinformerfactory, but if another informer consumer existed and was missed (or was added later without realizing this behavior was enabled), the failure mode is quite poor: an error occurs, the informer is stopped, but that consumer becomes a zombie, expecting events from a stopped informer.

These distinct capabilities would be easier to reason about in isolation:
1. Ability for someone that added an event handler to an informer to remove it (@kevindelgado mentioned there was a separate PR adding this)
    * GC and Quota controller loops already know they want to stop a previously started resource monitor (based on the resource being missing in discovery). It would be more reliable for them to remove their event handlers at that point, instead of relying on an independent top-level policy to stop informers when encountering errors. We don't actually want to stop GC and Quota informers on errors, we want to stop them when discovery indicates the resource types no longer exist.
    * Other controllers could use the ability to remove event handlers when shutting down, or for any other reason
2. Ability for informers to notice/stop once they have no event handlers registered
    * I think this is safer than a global "stop informer on error" policy, because it helps ensure we don't still have event handlers registered still expecting events, whose informers were actually silently stopped because of an error
    * I'm not sure what to do about handles to listers backed by informers... do we require they be "Close()"ed or ack stopping somehow as well?
3. Ability for an event handler to get notifications of list/watch errors
    * For informer consumers that *did* want to stop on error, this could be used to trigger removal of an event handler to make a local decision to stop in reaction to list/watch errors.

yliaog commented:
agree with @liggitt comments above, just want to quickly point out that this PR provides informer consumer a way to get notified when informer is stopped, hence informer consumer has a chance to react to the notification, instead of silently waiting for events from an already stopped informer.

kevindelgado commented:
I've done a major overhaul of this in response to https://github.com/kubernetes/kubernetes/pull/97214#issuecomment-810609141 

The only stop option offered now is when an informer has zero handlers registered. Note that the first commit is a dupe of https://github.com/kubernetes/kubernetes/pull/98657 which has already been LGTM'd and should land shortly.

PTAL @liggitt

kevindelgado commented:
/retest

kevindelgado committed:
Add OnError to ResourceEventHandler

kevindelgado committed:
Enable running SharedInformer with stop options

kevindelgado committed:
Use RunWithStopOptions in dynamic and metadata informer factories

kevindelgado committed:
Modify quota controller to stop informers on error

kevindelgado committed:
Controller Manager runs stoppable informers

kevindelgado commented:
Incorporated all feedback from https://github.com/kubernetes/kubernetes/pull/97214#issuecomment-810609141 and introduced new `OnError` method to the `ResourceEventHandler` interface based on the conversation in https://github.com/kevindelgado/kubernetes/pull/2

PTAL @yliaog and @liggitt

kevindelgado committed:
Factor out common CRD integration testing utils

kevindelgado committed:
Modify GC controller to stop informers on error

kevindelgado commented:
/retest

kevindelgado commented:
/retest

kevindelgado commented:
/retest

caesarxuchao commented:
/unassign

kevindelgado commented:
friendly ping @yliaog @liggitt

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -150,6 +152,15 @@ type SharedInformer interface {
+	// results in an appropriate error.
+	// Please note: If, for some reason, the same handler has been added multiple
+	// times, all registrations will be removed.
+	RemoveEventHandler(handler ResourceEventHandler) error
```
hmm ok, it looks like it will be changing back to returning an error, I will wait until https://github.com/kubernetes/kubernetes/pull/98657 is updated to sync with latest

atoato88 commented:
FYI
I confirmed that this PR clears off controller manager logs said in #79610 on my local env.
Thank you to create this PR.
I'll wait for merging to master.

kevindelgado commented:
> FYI
> I confirmed that this PR clears off controller manager logs said in #79610 on my local env.
> Thank you to create this PR.
> I'll wait for merging to master.

Yea sorry for how slow this is moving. This PR is blocked on https://github.com/kubernetes/kubernetes/pull/98657 which is still in review and I have been trying to push it forward without much luck. Feel free to ping that PR like I have been doing to try to see if progress can be made there.

atoato88 commented:
/remove-lifecycle stale

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

atoato88 commented:
I think this PR is still needed because #79610 error occurs on v1.23.0 cluster created by kind.

@kevindelgado 
Any comments and could you rebase?

kevindelgado commented:
This PR is blocked on https://github.com/kubernetes/kubernetes/pull/98657

The issues from #79610 are certainly still present and will need to be addressed, but this PR has been deprioritized given the current state of #98657

dims commented:
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close

Pingan2017 commented:
@kevindelgado 
we encountered a bug when we merge this PR and #98657 .
when delete a deployment, the gc controller set the `deletionTimestamp` for the rs and pod, then pod been deleted from etcd by kubelet, but the gc controller not delete rs and deploy form etcd.  
When we analyzed the problem, we found apiserver restarted when delete deployment and the pod still exist in gc controller cache even pod deleted from etcd. the gc controller miss the pod deleted event .
By reading the code, we found that this bug is related to this PR.
Let's imagine a scenario:
1. gc controller start informer and add event handle
2. deploy a deployment , gc controller will received `deployment`, `rs` , `pod` event , and add them to cache.
3. delete this `deployment`, gc controler set the `deletionTimestamp` for  `rs` , `pod` 
4. apiserver restart
5. gc controller sync the deleted
6. 
7. 
8.

oomichi commented:
> This PR is blocked on #98657
> 
> The issues from #79610 are certainly still present and will need to be addressed, but this PR has been deprioritized given the current state of #98657

Hi @kevindelgado 

https://github.com/kubernetes/kubernetes/pull/111122 has been merged instead of #98657
Can we move forward for this again?
